File: akka-javasdk/src/main/java/akka/javasdk/keyvalueentity/KeyValueEntityProvider.java
/*
 * Copyright (C) 2021-2024 Lightbend Inc. <https://www.lightbend.com>
 */

package akka.javasdk.keyvalueentity;

import akka.javasdk.impl.Kalix;
import akka.javasdk.impl.MessageCodec;
import com.google.protobuf.Descriptors;
import akka.javasdk.impl.keyvalueentity.KeyValueEntityRouter;

import java.util.Optional;

/**
 * Register a value based entity in {@link Kalix} using a <code>
 * KeyValueEntityProvider</code>. The concrete <code>KeyValueEntityProvider</code> is generated for the
 * specific entities defined in Protobuf, for example <code>CustomerEntityProvider</code>.
 */
public interface KeyValueEntityProvider<S, E extends KeyValueEntity<S>> {

  KeyValueEntityOptions options();

  Descriptors.ServiceDescriptor serviceDescriptor();

  String typeId();

  KeyValueEntityRouter<S, E> newRouter(KeyValueEntityContext context);

  Descriptors.FileDescriptor[] additionalDescriptors();

  default Optional<MessageCodec> alternativeCodec() {
    return Optional.empty();
  }
}

File: akka-javasdk/src/main/scala/akka/javasdk/impl/keyvalueentity/KeyValueEntityProvider.scala
/*
 * Copyright (C) 2021-2024 Lightbend Inc. <https://www.lightbend.com>
 */

package akka.javasdk.impl.keyvalueentity

import akka.annotation.InternalApi
import akka.javasdk.impl.ComponentDescriptor
import akka.javasdk.impl.ComponentDescriptorFactory
import akka.javasdk.impl.JsonMessageCodec
import akka.javasdk.keyvalueentity.KeyValueEntity
import akka.javasdk.keyvalueentity.KeyValueEntityContext
import com.google.protobuf.Descriptors

import java.util.function.Function

/**
 * INTERNAL API
 */
@InternalApi
private[impl] final case class KeyValueEntityProvider[S, E <: KeyValueEntity[S]](
    entityClass: Class[E],
    messageCodec: JsonMessageCodec,
    factory: Function[KeyValueEntityContext, E]) {
  private val typeId: String = ComponentDescriptorFactory.readComponentIdIdValue(entityClass)
  if (typeId eq null)
    throw new IllegalArgumentException(
      "Key Value Entity [" + entityClass.getName + "] is missing '@ComponentId' annotation")
  private val componentDescriptor = ComponentDescriptor.descriptorFor(entityClass, messageCodec)

  val serviceDescriptor: Descriptors.ServiceDescriptor = componentDescriptor.serviceDescriptor

  private def newRouter(context: KeyValueEntityContext): KeyValueEntityRouter[S, E] = {
    val entity = factory.apply(context)
    new ReflectiveKeyValueEntityRouter[S, E](entity, componentDescriptor.commandHandlers)
  }

  def newServiceInstance(): KeyValueEntityService = {
    new KeyValueEntityService(
      newRouter,
      serviceDescriptor,
      Array[Descriptors.FileDescriptor](componentDescriptor.fileDescriptor),
      messageCodec,
      typeId)
  }

}
